rfcs/0084-partial-checkpoint.md

```diff
@@ -0,0 +1,162 @@
+    
+    Therefore, in a checkpoint that can serve batch query, states do not need to have an aligned max committed epoch(MCE), and **each state can have its own MCE**. 
+    
+    By default, a batch query will use the **minimum MCE among all states** as the read epoch. This is the **same behavior as the current system**. If users are interested in some subset of states, with some ways of proper setting (user interface TBD), it will use the minimum MCE among this subset of states as the read epoch.
+    
+    In conclusion, for batch query, in a checkpoint we don’t have to have an global aligned MCE among all states.
+    
+
+In a word, in the proposed design
+
+- in a checkpoint, states will not have a global aligned MCE, and instead each state may have their own MCE. The MCE of upstream is always higher than their downstreams.
+- During failure recovery, the system can get back to work immediately from a checkpoint. For a downstream MV that reads from upstream, if it lags behind, it will first replay the change log of the lagged epochs of the upstream MV (barriers of the lagged epochs are also replayed), and then process the new data from the upstream.
+
+## Detailed Design
+
+### Partial checkpoint with unaligned MCE
+
+> Copied from the previous design doc. 
+**snapshot ⇒ checkpoint**
+> 
+
+Let’s look at the following graph. The MCE of each MView is e2 at the beginning and we have a snapshot 1
+
+![](images/0084-partial-checkpoint/partial1.png)
+
+When all actors in the subgraph of MView1 has received the barrier and synced its state, we commit e3 only on MView1 and create a new snapshot 2
+
+![](images/0084-partial-checkpoint/partial2.png)
+
+When all actors in the subgraph of MView2 has received the barrier and synced its state, we commit e3 on MView2 and create a new snapshot 3.
+
+![](images/0084-partial-checkpoint/partial3.png)
+
+### L0 as a log
+
+Implementation details in [Log Data Management](https://github.com/risingwavelabs/rfcs/pull/55) 
+
+Summary:
+
+- No extra storage. Reuse the L0 data written by upstream MVs. Materialized state writes extra old value.
+- Easy truncation. Only metadata change.
+
+### Introducing the New Executors to Read Upstream MV
+
+Currently on MV-on-MV, we will forward the new data flowed from upstream directly to downstream backfill executor.
+
+![](images/0084-partial-checkpoint/current-chain-exec.png)
+
+Under the new design, if a MV lags behind the upstream MV, it will first replay the change log of the lagged epoch of the upstream, and then send the new data in upstream to downstream executor. Following is an example.
```

Comment:
So all MVs will have changelog store?
Is it created per MV? Or all MVs will share one per upstream.

```
upstream -> log store -> mv
                -> log store -> mv
```

Or
```
upstream -> log store -> mv
                                 \
                                   \ - > mv
```

Reply:
The later one
```
upstream -> log store -> mv
                                 \
                                   \ - > mv
```